x-pack/plugins/security_solution/server/endpoint/lib/policy/license_watch.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Subscription } from 'rxjs';
+
+import {
+  KibanaRequest,
+  Logger,
+  SavedObjectsClientContract,
+  SavedObjectsServiceStart,
+} from 'src/core/server';
+import { PackagePolicy, PACKAGE_POLICY_SAVED_OBJECT_TYPE } from '../../../../../fleet/common';
+import { PackagePolicyServiceInterface } from '../../../../../fleet/server';
+import { ILicense } from '../../../../../licensing/common/types';
+import {
+  isEndpointPolicyValidForLicense,
+  unsetPolicyFeaturesAboveLicenseLevel,
+} from '../../../../common/license/policy_config';
+import { licenseService } from '../../../lib/license/license';
+
+export class PolicyWatcher {
+  private logger: Logger;
+  private soClient: SavedObjectsClientContract;
+  private policyService: PackagePolicyServiceInterface;
+  private subscription: Subscription | undefined;
+  constructor(
+    policyService: PackagePolicyServiceInterface,
+    soStart: SavedObjectsServiceStart,
+    logger: Logger
+  ) {
+    this.policyService = policyService;
+    this.soClient = this.makeInternalSOClient(soStart);
+    this.logger = logger;
+  }
+
+  /**
+   * The policy watcher is not called as part of a HTTP request chain, where the
+   * request-scoped SOClient could be passed down. It is called via license observable
+   * changes. We are acting as the 'system' in response to license changes, so we are
+   * intentionally using the system user here. Be very aware of what you are using this
+   * client to do
+   */
+  private makeInternalSOClient(soStart: SavedObjectsServiceStart): SavedObjectsClientContract {
+    const fakeRequest = ({
+      headers: {},
+      getBasePath: () => '',
+      path: '/',
+      route: { settings: {} },
+      url: { href: {} },
+      raw: { req: { url: '/' } },
+    } as unknown) as KibanaRequest;
+    return soStart.getScopedClient(fakeRequest, { excludedWrappers: ['security'] });
+  }
+
+  public start() {
+    this.subscription = licenseService.getLicenseInformation$()?.subscribe(this.watch.bind(this));
+  }
+
+  public stop() {
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+
+  public async watch(license: ILicense) {
```

Comment:
At one point we had talked about possibly doing the "migration" slightly delayed from when the observable triggers in order to try and account for edge case where requests might be in flight that would override our changes here.  Delaying the migration would allow for existing one to complete ++ any future ones would be auto-rejected by the API handles, so migrating should be able to account for all records.

Also - do you need to protect against the license changing while a migration is in flight?

Reply:
I don't think I was part of any of these discussions. We want to add a sleep here?

Do we expect other things changing at the same moment the license changes? What's this about migrations?

Reply:
Re: 
> Do we expect other things changing at the same moment the license changes? What's this about migrations?

My understanding is that changing of license can happen while the system is running - there is downtime when it happens. It just happens


re: 
> What's this about migrations?

Sorry for the confusion. In my mind, what this service here is doing is a migration - thats what I mean.